File: design_patterns/C++/Creational/builder.cxx
class BurgerBuilder;

class Burger {
  protected:
    int size = 0;
    bool cheese = false;
    bool pepperoni = false;
    bool lettuce = false;
    bool tomato = false;

  public:
    Burger(BurgerBuilder* builder);
};

class BurgerBuilder {
  public:
    int size = 0;
    bool cheese = false;
    bool pepperoni = false;
    bool lettuce = false;
    bool tomato = false;

    BurgerBuilder(int size) {
        this->size = size;
    }

    BurgerBuilder* addPepperoni() {
        this->pepperoni = true;
        return this;
    }

    BurgerBuilder* addCheese() {
        this->cheese = true;
        return this;
    }

    BurgerBuilder* addLettuce() {
        this->lettuce = true;
        return this;
    }

    BurgerBuilder* addTomato() {
        this->tomato = true;
        return this;
    }

    Burger* build() {
        return new Burger(this);
    }
};

Burger::Burger(BurgerBuilder* builder) {
    this->size = builder->size;
    this->cheese = builder->cheese;
    this->pepperoni = builder->pepperoni;
    this->lettuce = builder->lettuce;
    this->tomato = builder->tomato;
}

int main() {
    auto burger = (new BurgerBuilder(6))
                  ->addPepperoni()
                  ->addLettuce()
                  ->addTomato()
                  ->build();
    return 0;
}

File: design_patterns/Python/Creational/builder.py
class Burger:

    def __init__(self, builder):
        self.size = builder.size
        self.cheese = builder.cheese
        self.pepperoni = builder.pepperoni
        self.lettuce = builder.lettuce
        self.tomato = builder.tomato


class BurgerBuilder:

    def __init__(self, size):
        self.size = size
        self.cheese = False
        self.pepperoni = False
        self.lettuce = False
        self.tomato = False

    def add_cheese(self):
        self.cheese = True

    def add_pepperoni(self):
        self.pepperoni = True

    def add_lettuce(self):
        self.lettuce = True

    def add_tomato(self):
        self.tomato = True

    def build(self):
        return Burger(self)


if __name__ == '__main__':

    burger = BurgerBuilder(6)
    burger.add_cheese()
    burger.add_lettuce()
    burger.build()
